tokio/src/task/join_map.rs

```diff
@@ -0,0 +1,782 @@
+            jh.abort();
+        }
+    }
+
+    /// Waits until one of the tasks in the map completes and returns its
+    /// output, along with the key corresponding to that task.
+    ///
+    /// Returns `None` if the map is empty.
+    ///
+    /// # Cancel Safety
+    ///
+    /// This method is cancel safe. If `join_one` is used as the event in a `tokio::select!`
+    /// statement and some other branch completes first, it is guaranteed that no tasks were
+    /// removed from this `JoinMap`.
+    ///
+    /// # Returns
+    ///
+    /// This function returns:
+    ///
+    ///  * `Some((key, Ok(value)))` if one of the tasks in this `JoinMap` has
+    ///    completed. The `value` is the return value of that ask, and `key` is
+    ///    the key associated with the task.
+    ///  * `Some((key, Err(err))` if one of the tasks in this JoinMap` has
+    ///    panicked or been aborted. `key` is the key associated  with the task
+    ///    that panicked or was aborted.
+    ///  * `None` if the `JoinMap` is empty.
+    pub async fn join_one(&mut self) -> Option<(K, Result<V, JoinError>)> {
+        crate::future::poll_fn(|cx| self.poll_join_one(cx)).await
+    }
+
+    /// keys all tasks and waits for them to finish shutting down.
+    ///
+    /// Calling this method is equivalent to calling [`abort_all`] and then calling [`join_one`] in
+    /// a loop until it returns `None`.
+    ///
+    /// This method ignores any panics in the tasks shutting down. When this call returns, the
+    /// `JoinMap` will be empty.
+    ///
+    /// [`abort_all`]: fn@Self::abort_all
+    /// [`join_one`]: fn@Self::join_one
+    pub async fn shutdown(&mut self) {
+        self.abort_all();
+        while self.join_one().await.is_some() {}
+    }
+
+    /// Abort the task corresponding to the provided `key`.
+    ///
+    /// If this `JoinMap` contains a task corresponding to `key`, this method
+    /// will abort that task and return `true`. Otherwise, if no task exists for
+    /// `key`, this method returns `false`.
+    pub fn abort<Q: ?Sized>(&mut self, key: &Q) -> bool
+    where
+        Q: Hash + Eq,
+        K: Borrow<Q>,
+    {
+        let key = self.pre_hashed_key(key);
+        let task_set = &mut self.task_set;
+        let entry = match self.key_set.get(&key as &dyn DynKey<Q>) {
+            Some(entry) => entry,
+            None => return false,
+        };
+
+        if let Some(mut entry) = task_set.entry_mut(entry.task.entry.clone()) {
+            entry.with_value_and_context(|(_actual_hash, jh), _| {
+                jh.abort();
+            });
+            return true;
+        }
+
+        false
+    }
+
+    /// keys all tasks with keys matching `predicate`.
+    ///
+    /// `predicate` is a function called with a reference to each key in the
+    /// map. If it returns `true` for a given key, the corresponding task will
+    /// be cancelled.
+    // XXX(eliza): do we want to consider counting the number of tasks aborted?
```

Comment:
We should probably clean this comment up a bit

Reply:
just to clarify, do you just mean removing the `// XXX` comment, or are you suggesting changes to the RustDoc comment as well?

Reply:
The XXX

Reply:
This was intended for PR reviewers and to be removed before merge if I understand correctly?

Reply:
correct. after thinking about it a bit, I don't think it's really necessary for us to do this, as the user-provided predicate can always count aborted tasks if desired...